adding additional functionality to 'add_selectors'

edmoss345 opened the pull request:
@fagiothree think I have added the additional functionality that we discussed, please test on your end. Have updated the README so hopefully makes sense

closes #35

istride reviewed (commented):
Would you let me know how I should test this script? Which input parameters can I use?

You need be careful with taking input from the command line. All command line arguments come into the script as strings, which you will then need to parse to obtain any other kind of data type. If the args processing is getting quite sophisticated then I would consider introducing a purpose-built library like [yargs] to help out.

Alternatively, it may be easiest to read `special_words` from a file - one word per line. It might be best to keep matching `add_selectors` with the string value "yes", as before.

[yargs]: https://yargs.js.org/

edmoss345 commented:
Thanks for the feedback @istride , good point on the 'special_words' input, think the script is working at the moment but as you say it is just being handled as a long string rather than a list. I've added some files which you can use to test functionality. Not sure these should stay in but should allow you to run some tests from your end then maybe we can write some very specific tests for the functions. For running the script use something like:

node index move_quick_replies "test/Test_Files/Before_qr_to_msg_text.json" "test/Test_Files/select_phrases.json" <OUTPUT NAME> <OUTPUT DIR> "['back', 'other']" true

I've also added some ideal before and after screenshots from RapidPro so you can see what the script is meant to be doing. We are moving quick replies to message text, adding numbers in their place (this is all existing script), but where we have 'special_words' we do not add numbers in but instead keep the original quick reply

edmoss345 commented:
@istride  have added some testing as part of this PR, could you have a look to see if what you think I've done is ok and we can discuss further in our catch up

edmoss345 commented:
@fagiothree I think it's up to you but yes you're probably right, it makes sense to store them in a json file similar to "select phrases", I'll make that change.

Sorry I don't think we've ever talked about translation, what we have at the moment will only work on the English. Both scenarios you have mentioned above are possible, either we can just take the appropriate translation from the localization, or we could have the requirement to add the translated words into the "special_words" JSON. What would your preference be? The functionality of the two options would be slightly different as if we take from localization the behaviour will be the same across all languages, if we input all the words into "special_words" then we could have different behaviour for different languages. Eg. If the word "back" is super long in another language we could choose not to add it back in its full form and just stick with the numbers.

edmoss345 commented:
@fagiothree , any further thoughts on the above, how you would like the code to handle the translations?

fagiothree commented:
Sorry for the late reply. I think I would go for the option of specifying
also the translation in the list of special words, since it allows more
flexibility and also includes the first option as a subcase. What do you
think?

On Tue, Apr 4, 2023 at 8:15 PM Ed Moss ***@***.***> wrote:

> @fagiothree <https://github.com/fagiothree> , any further thoughts on the
> above, how you would like the code to handle the translations?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/IDEMSInternational/idems_translation/pull/30#issuecomment-1496400046>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHX4SUEOZBPNGQNYS4IQJX3W7RQN3ANCNFSM6AAAAAAVNQUKZY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

edmoss345 commented:
Agreed @fagiothree , I think that is more flexible, I will implement as suggested

edmoss345 commented:
Closes #35

edmoss345 commented:
@fagiothree think I have fixed the problems that you flagged earlier. As part of this process I also did a bit of a refactor to make the code a bit easier to read through. Could you do the check you were doing before, if you are happy things are working could you ping @istride to have a review as well.

@istride , I think it is a bit less of a mess than before but still not super clean. When Chiara gives the ok could you have a general look for any glaring errors. We can discuss more major refactors in due course.

fagiothree commented:
@istride I tested the code and it's working fine for me, so could have a general look as well before we merge? Thank you!

istride commented on chatbot/index.js:
```diff
@@ -80,9 +80,14 @@ function localize([inputFlow, translations, lang, outputName, outputDir]) {
 
 
-function move_quick_replies([input_file, select_phrases, outputName, outputDir,add_selectors = false]) {
+function move_quick_replies([input_file, select_phrases, outputName, outputDir, add_selectors = "no", special_words = false]) {
```
It's not possible for more than 1 argument to be optional. If I don't specify a value for 'add selectors' (because the default is "no"), but specify a path for 'special words', then `special_words` will be `undefined` and `add_selectors` will be the path to the special words file.

I suggest making add_selectors a required argument - it must be either yes/no (or true/false, whatever you decide to go with), but it cannot be left out. The only truly optional argument is the last one in the list.

In future we should consider using a CLI arguments parsing library like 'yargs', if the arguments continue to get more sophisticated.

edmoss345 replied:
Understood, will make it a required argument "yes/no"

istride commented on chatbot/insert/add_quick_replies_to_msg_text_and_localization.js:
```diff
@@ -1,4 +1,14 @@
-function move_quick_replies_to_message_text(flows, select_phrases,add_selectors) {
+// const fs = require('fs');
+// [flows, debug, debug_lang] = move_quick_replies_to_message_text(
+//         readInputFile("./chatbot/test/Input/Tester.json"),readInputFile("./chatbot/test/Input/select_phrases.json"), "yes", readInputFile("./chatbot/test/Input/special_words.json")
+// )
+// function readInputFile(filePath) {
+//     return JSON.parse(fs.readFileSync(filePath).toString());
+// }
```
Delete commented code, if not needed.

edmoss345 replied:
deleted

istride commented on chatbot/insert/add_quick_replies_to_msg_text_and_localization.js:
```diff
@@ -131,271 +168,145 @@ function modify_router_node_cases(flow, node, action, curr_loc, quick_replies, r
 
     let router = routers[dest_id];
     if (router) {
+        // loop through all the cases in the wfr node
```
What is 'wfr'?

It is evident from the code that we are looping through all the cases, so I'm not sure that this comment is useful.

istride replied:
Reading further, I can see that there are many comments like this that seem to be merely stating what the code is doing. Please review these comments, if you have added them, and think about whether they are actually going to help a future reader of the code to understand it better.

My opinion on comments is that they should be avoided, and only used as a last resort, for example, to explain _why_ the code was written in this particular way (if it might be surprising), or to give some context. If you feel the need to explain what the code does, then that is a strong suggestion that the code may need to be refactored so that it's easier to read and understand.

edmoss345 replied:
Understood, have gone through and taken out the unnecessary comments

istride commented on chatbot/test/fix-arg-qr-translation-test.js:
```diff
@@ -0,0 +1,23 @@
+    return JSON.parse(fs.readFileSync(filePath).toString());
+}
+
+describe('Test of "Fixer Arg and QR after translation" function', function() {
```
I'm not a fan of this "Test of 'x' function" boilerplate. I think you can just keep the "x" bit and drop everything else.

istride commented on chatbot/test/fix-arg-qr-translation-test.js:
```diff
@@ -0,0 +1,23 @@
+}
+
+describe('Test of "Fixer Arg and QR after translation" function', function() {
+    it('should fix the arguments of various nodes to ensure the logic of the english is consistent with the translation, runs the function on a sample input and check it matches the ideal output file', function() {
```
I like long-ish descriptions of tests, but these should still be kept succinct. "runs the function on a sample input and check it matches the ideal output file" is unnecessary and should be removed.

edmoss345 replied:
Changed here and elsewhere
